pageserver/src/tenant.rs

```diff
@@ -1350,13 +1268,64 @@ impl Tenant {
                         .map_err(LoadLocalTimelineError::ResumeDeletion);
                     }
 
-                    // We're loading fresh timeline that didnt yet make it into remote.
-                    (None, Some(remote_client))
+                    // see if there is at most one layer file and a metadata file in the directory
+                    // if not, we should fail the loading
+                    //
+                    // otherwise delete the not yet uploaded timeline.
+                    //
+                    // this timeline cannot have children because those are not allowed to be
+                    // created before uploading.
+
+                    let path = self.conf.timeline_path(&self.tenant_id, &timeline_id);
+
+                    return tokio::task::spawn_blocking({
+                        let path = path.clone();
+                        move || {
+                            let mut metadata = false;
+                            let mut layers = 0;
+                            let mut others = 0;
+                            for dentry in path.read_dir_utf8()? {
+                                let dentry = dentry?;
+                                let file_name = dentry.file_name();
+
+                                if file_name == "metadata" {
+                                    metadata = true;
+                                    continue;
+                                }
+
+                                use std::str::FromStr;
+
+                                if crate::tenant::storage_layer::LayerFileName::from_str(file_name)
+                                    .is_ok()
+                                {
+                                    layers += 1;
+                                    continue;
+                                }
+
+                                others += 1;
+                            }
+
+                            // bootstrapped have the one image layer file, or one partial temp file
+                            if !(metadata && layers + others <= 1) {
+                                anyhow::bail!("unexpected assumed fresh timeline: found metadata={}, layers={}, others={}", metadata, layers, others);
+                            }
+
+                            let tmp_path =
+                                path.with_file_name(format!("{timeline_id}{}", TEMP_FILE_SUFFIX));
+                            std::fs::rename(path, &tmp_path)?;
+                            std::fs::remove_dir_all(&tmp_path)?;
```

Comment:
Aim is to do the similar trick as "safe delete tenant dir" does, but without tempdir reimplementation. Idea was that the `__temp` suffix would be deleted in startup, however I wrote this when I was still thinking that `TimelineId::from_str` would accept trailing garbage.

Need to check this.

Reply:
It is accepted as a temporary timeline as the check is just for suffix.